### PR TITLE
feat: Add Hyperpod Optimum-neuron LoRA example

### PR DESCRIPTION
**Pull request by Captainia**
*Issue #, if available:* N/A, new feature 

*Description of changes:*

This PR adds an example to the test_cases, by using Huggingface optimum-neuron library for PEFT fine-tuning. This repo is is the interface between the Huggingface Transformers library and AWS Accelerators including [AWS Trainium](https://aws.amazon.com/machine-learning/trainium/?nc1=h_ls) and [AWS Inferentia](https://aws.amazon.com/machine-learning/inferentia/?nc1=h_ls). It provides a set of tools enabling easy model loading, training and inference on single- and multi-Accelerator settings for different downstream tasks. 

The example uses Hyperpod EKS environment, and the code will used in the Hyperpod workshop 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by mhuguesaws — CHANGES_REQUESTED** [review]
Thank you for your contribution.

Could you please:
1. join the AWS organization on github
2. organize the folder as `pytorch/fine-tuning/optimum-neuron`
3. Add the kubernetes manifest into a kubernetes folder. The repo structure organizes by scheduler.

**Comment by Captainia** [comment]
> Thank you for your contribution.
> 
> Could you please:
> 
> 1. join the AWS organization on github
> 2. organize the folder as `pytorch/fine-tuning/optimum-neuron`
> 3. Add the kubernetes manifest into a kubernetes folder. The repo structure organizes by scheduler.

Thank you for the quick review!
1. I am in the AWS organization, https://github.com/aws and https://github.com/awslabs, is there another one I should join?
2,3. There will be another continuous pre-training example under `pytorch/optimum-neuron/llama3/kubernetes/continuous-pre-training`, so I organized currently as `pytorch/optimum-neuron/llama3/kubernetes/fine-tuning`. Following `neuronx-distributed` Does this sounds good?

**Comment by mhuguesaws** [comment]
> > Thank you for your contribution.
> > Could you please:
> > 
> > 1. join the AWS organization on github
> > 2. organize the folder as `pytorch/fine-tuning/optimum-neuron`
> > 3. Add the kubernetes manifest into a kubernetes folder. The repo structure organizes by scheduler.
> 
> Thank you for the quick review!
> 
> 1. I am in the AWS organization, https://github.com/aws and https://github.com/awslabs, is there another one I should join?
>    2,3. There will be another continuous pre-training example under `pytorch/optimum-neuron/llama3/kubernetes/continuous-pre-training`, so I organized currently as `pytorch/optimum-neuron/llama3/kubernetes/fine-tuning`. Following `neuronx-distributed` Does this sounds good?

aws-samples organization

2and 3 are good.

File: 3.test_cases/pytorch/optimum-neuron/llama3/kubernetes/fine-tuning/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+# # Set pip repository pointing to the Neuron repository 
+RUN python -m pip config set global.extra-index-url https://pip.repos.neuron.amazonaws.com
+
+# # Install wget, awscli 
+RUN python -m pip install wget 
+RUN python -m pip install awscli 
+
+# # Update Neuron Compiler and Framework
+RUN python -m pip install --upgrade neuronx-cc==2.* torch-neuronx==2.1.* torchvision
+RUN python -m pip install --upgrade neuronx-distributed neuronx-distributed-training
```
Comment (jianyinglangaws) [inline 3.test_cases/pytorch/optimum-neuron/llama3/kubernetes/fine-tuning/Dockerfile:27]: Can we fix the library version? A new release could break the compatibility.

File: 3.test_cases/pytorch/optimum-neuron/llama3/kubernetes/fine-tuning/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+RUN mkdir workspace
+
+COPY src/peft_tokenize_data.py workspace/peft_tokenize_data.py
+COPY src/peft_train.py workspace/peft_train.py
+COPY src/consolidation.py workspace/consolidation.py
+COPY src/merge_lora.py workspace/merge_lora.py
+
+RUN python3 -m pip install datasets tokenizers peft==0.14.0 huggingface_hub trl==0.11.4 PyYAML accelerate==0.29.2
+
+RUN pip install git+https://github.com/huggingface/optimum-neuron.git
```
Comment (jianyinglangaws) [inline 3.test_cases/pytorch/optimum-neuron/llama3/kubernetes/fine-tuning/Dockerfile:38]: Fix the version

File: 3.test_cases/pytorch/optimum-neuron/llama3/kubernetes/fine-tuning/src/peft_train.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
+
```
Comment (KeitaW) [inline 3.test_cases/pytorch/optimum-neuron/llama3/kubernetes/fine-tuning/src/peft_train.py:1]: Could we remove those *.py files and refer original scripts instead? ex. https://github.com/huggingface/optimum-neuron/blob/1fd77f2ed1da434418ec53b67365b336dc40c57a/docs/source/training_tutorials/sft_lora_finetune_llm.py#L4 We cannot maintain those training code on the repo.

**Review by KeitaW — CHANGES_REQUESTED** [review]
Left a comment

File: 3.test_cases/pytorch/optimum-neuron/llama3/kubernetes/fine-tuning/src/peft_train.py (hunk 3 above)
Comment (KeitaW) [reply 3.test_cases/pytorch/optimum-neuron/llama3/kubernetes/fine-tuning/src/peft_train.py:1]: Same for each *.py file, if there's original

File: 3.test_cases/pytorch/optimum-neuron/llama3/kubernetes/fine-tuning/README.md [hunk 4]
```diff
@@ -0,0 +1,148 @@
+## PEFT fine tuning of Llama 3 on SageMaker HyperPod with AWS Trainium
```
Comment (KeitaW) [inline 3.test_cases/pytorch/optimum-neuron/llama3/kubernetes/fine-tuning/README.md:1]:
```suggestion
## PEFT fine tuning of Llama 3 on Amazon EKS with AWS Trainium
```

Nothing HP specific

File: 3.test_cases/pytorch/optimum-neuron/llama3/kubernetes/fine-tuning/README.md [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+This README demonstrates how to perform efficient supervised fine tuning for a Meta Llama 3.1 model using Parameter-Efficient Fine Tuning (PEFT) on AWS Trainium with SageMaker HyperPod. We use HuggingFace's Optimum-Neuron SDK to apply Low-Rank Adaptation (LoRA) to fine-tuning jobs, and use SageMaker HyperPod as the primary compute cluster to perform distributed training on Trainium.
```
Comment (KeitaW) [inline 3.test_cases/pytorch/optimum-neuron/llama3/kubernetes/fine-tuning/README.md:2]:
```suggestion
This README demonstrates how to perform efficient supervised fine tuning for a Meta Llama 3.1 model using Parameter-Efficient Fine Tuning (PEFT) on AWS Trainium with EKS. We use HuggingFace's Optimum-Neuron SDK to apply Low-Rank Adaptation (LoRA) to fine-tuning jobs, and use Trainiums to perform distributed trainin.
```

File: 3.test_cases/pytorch/optimum-neuron/llama3/kubernetes/fine-tuning/README.md [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+## 0. Prerequisites
+
+### 0.1. EKS Cluster 
+Before running this training, you'll need to create an Amazon EKS or a SageMaker HyperPod EKS cluster with at least 1 trn1.32xlarge/ trn1n.32xlarge. Instructions can be found in [1.architectures](../../1.architectures), the [aws-do-eks](https://bit.ly/do-eks) project, or the [eks-blueprints](https://github.com/aws-ia/terraform-aws-eks-blueprints) project.
+
+### 0.2 Setup Persistant Volume Claim(PVC) for fsx 
+
+We need to setup an PVC for FSx to store the tokenized data and training checkpoints. Please follow the link [here](https://catalog.workshops.aws/sagemaker-hyperpod-eks/en-US/01-cluster/06-fsx-for-lustre) to setup FSx CSI Driver and PVC. 
+
```
Comment (KeitaW) [inline 3.test_cases/pytorch/optimum-neuron/llama3/kubernetes/fine-tuning/README.md:20]: HF_TOKEN access should be in the prerequisites.

File: 3.test_cases/pytorch/optimum-neuron/llama3/kubernetes/fine-tuning/README.md [hunk 7]
```diff
@@ -0,0 +1,148 @@
+## PEFT fine tuning of Llama 3 on SageMaker HyperPod with AWS Trainium
+This README demonstrates how to perform efficient supervised fine tuning for a Meta Llama 3.1 model using Parameter-Efficient Fine Tuning (PEFT) on AWS Trainium with SageMaker HyperPod. We use HuggingFace's Optimum-Neuron SDK to apply Low-Rank Adaptation (LoRA) to fine-tuning jobs, and use SageMaker HyperPod as the primary compute cluster to perform distributed training on Trainium.
+
+### Solution overview
+This solution uses the following components:
+
+### SageMaker HyperPod for distributed training infrastructure
```
Comment (KeitaW) [inline 3.test_cases/pytorch/optimum-neuron/llama3/kubernetes/fine-tuning/README.md:7]:
```suggestion
### Distributed training infrastructure
```

File: 3.test_cases/pytorch/optimum-neuron/llama3/kubernetes/fine-tuning/src/peft_train.py (hunk 3 above)
Comment (Captainia) [reply 3.test_cases/pytorch/optimum-neuron/llama3/kubernetes/fine-tuning/src/peft_train.py:1]: Thanks for the suggestion @KeitaW , I agree with this from a high level to reduce code to maintain, however in particularly when referencing third-party dependency like optimum-neuron I think it is safer to maintain the code here due to several reasons:
1. We do have some changes here that the script do not support, in particularly caching the model artifact and saving tokenizer following the other examples. The optimum-neuron script will always download the tokenizer on the fly, which causes long time for the workshop on each run. 
2. We are also using a light weight lora config compared to original one, making the workshop focus on Amazon EKS for faster iteration.
3. We hear there is no guarantee that the examples will stay at the (path)[https://github.com/aws/deep-learning-containers/pull/4670/files#diff-0f776bad437279bcc3d6005ec1b29170f0b4e53dfbc5c3c234fb202817f707a3], thus it would be breaking for us if optimum-neuron changes the downstream example or they move it to a different place. 

I would suggest we keep this light weight script to make sure compatibility and availability, please let me know how you think about this. Thanks

File: 3.test_cases/pytorch/optimum-neuron/llama3/kubernetes/fine-tuning/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+RUN mkdir workspace
+
+COPY src/peft_tokenize_data.py workspace/peft_tokenize_data.py
+COPY src/peft_train.py workspace/peft_train.py
+COPY src/consolidation.py workspace/consolidation.py
+COPY src/merge_lora.py workspace/merge_lora.py
+
+RUN python3 -m pip install datasets==2.18.0 tokenizers==0.21.1 peft==0.14.0 huggingface_hub trl==0.11.4 PyYAML accelerate==0.29.2
+
+RUN pip install optumum-neuron==0.1.0
```
Comment (jianyinglangaws) [inline 3.test_cases/pytorch/optimum-neuron/llama3/kubernetes/fine-tuning/Dockerfile:38]: Typo `optumum-neuron` -> `optimum-neuron`

File: 3.test_cases/pytorch/optimum-neuron/llama3/kubernetes/fine-tuning/src/peft_train.py (hunk 3 above)
Comment (Captainia) [reply 3.test_cases/pytorch/optimum-neuron/llama3/kubernetes/fine-tuning/src/peft_train.py:1]: Yes the test case is for the workshop (Neuron x Hyperpod in 4/17 as planned), we have identified a gap that the Hyperpod cluster currently do not support trn1.2xlarge, while the workshop studio does not support trn1.32xlarge. Will update this PR as soon as there is a next step for this gap.

File: 3.test_cases/pytorch/optimum-neuron/llama3/kubernetes/fine-tuning/generate-jobspec.sh [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+export INSTANCE_TYPE=ml.trn1.32xlarge
+export EFA_PER_NODE=8
+export NEURON_PER_NODE=16
+
+# Storage Configuration
+export FSX_CLAIM=fsx-claim
+
+# Model and Dataset Configuration
+export HF_ACCESS_TOKEN=<Your Huggingface Token>
+export MODEL_ID=meta-llama/Llama-3.1-8B-Instruct
```
Comment (jianyinglangaws) [inline 3.test_cases/pytorch/optimum-neuron/llama3/kubernetes/fine-tuning/generate-jobspec.sh:22]: Can we change it to the in-gated model so people do not need to provide tokens?

Comment (Captainia) [reply 3.test_cases/pytorch/optimum-neuron/llama3/kubernetes/fine-tuning/generate-jobspec.sh]: Yes, I have updated to use the non-gated model.

**Review by KeitaW — APPROVED** [review]

**Review by jianyinglangaws — APPROVED** [review]
Looks good to me!

**Review by Captainia — COMMENTED** [review]
Thank you, the requested changes are committed.
